dmoj/cptbox/utils.py

```diff
@@ -1,11 +1,89 @@
+import errno
 import io
+import mmap
+import os
+from abc import ABCMeta, abstractmethod
+from tempfile import NamedTemporaryFile
+from typing import Optional
 
 from dmoj.cptbox._cptbox import memory_fd_create, memory_fd_seal
 
 
-class MemoryIO(io.FileIO):
-    def __init__(self) -> None:
-        super().__init__(memory_fd_create(), 'r+')
+def _make_fd_readonly(fd):
+    new_fd = os.open(f'/proc/self/fd/{fd}', os.O_RDONLY)
+    try:
+        os.dup2(new_fd, fd)
+    finally:
+        os.close(new_fd)
+
+
+class MmapableIO(io.FileIO, metaclass=ABCMeta):
+    def __init__(self, fd, *, prefill: Optional[bytes] = None, seal=False) -> None:
+        super().__init__(fd, 'r+')
+
+        if prefill:
+            self.write(prefill)
+        if seal:
+            self.seal()
+
+    @abstractmethod
+    def seal(self) -> None:
+        ...
+
+    def to_bytes(self) -> bytes:
+        try:
+            with mmap.mmap(self.fileno(), 0, access=mmap.ACCESS_READ) as f:
+                return bytes(f)
+        except ValueError as e:
+            if e.args[0] == 'cannot mmap an empty file':
+                return b''
+            raise
+
+
+class NamedFileIO(MmapableIO):
+    _name: str
+
+    def __init__(self, *, prefill: Optional[bytes] = None, seal=False) -> None:
+        with NamedTemporaryFile(delete=False) as f:
+            self._name = f.name
+            super().__init__(os.dup(f.fileno()), prefill=prefill, seal=seal)
+
+    def seal(self) -> None:
+        self.seek(0, os.SEEK_SET)
+
+    def close(self) -> None:
+        super().close()
+        os.unlink(self._name)
```

Comment:
Probably we should also implement `__del__`, since I'm not convinced we call `free_data` in all cases.

Reply:
It's implemented by default:

```pycon
Python 3.11.2 (main, Sep 14 2024, 03:00:30) [GCC 12.2.0] on linux
Type "help", "copyright", "credits" or "license" for more information.
>>> from dmoj.cptbox.utils import NamedFileIO
>>> NamedFileIO()
<_io.FileIO name=4 mode='rb+' closefd=True>
>>> 0
unlinking
0
```